src/mm/page_manager.rs

```diff
@@ -0,0 +1,214 @@
+use crate::kprintln;
+use crate::mm;
+use core::mem;
+use super::page_alloc::{PageAllocator, AllocError};
+
+#[derive(Debug, PartialEq)]
+pub enum PageKind {
+    /// For now, all reserved pages are owned by OpenSBI.
+    Reserved,
+    Kernel,
+    Used,
+    Unused,
+}
+
+/// Holds data about each physical page in the system.
+#[derive(Debug)]
+pub struct PhysicalPage {
+    kind: PageKind,
+    base: usize,
+
+    /// Page is last of a contiguous allocation of pages.
+    last: bool, 
+}
+
+impl PhysicalPage {
+    fn is_unused(&self) -> bool {
+        self.kind == PageKind::Unused
+    }
+
+    fn is_used(&self) -> bool {
+        self.kind != PageKind::Unused
+    }
+
+    fn set_used(&mut self) {
+        self.kind = PageKind::Used;
+    }
+
+    fn _is_last(&self) -> bool {
+        self.last
+    }
+
+    fn set_last(&mut self) {
+        self.last = true;
+    }
+
+    pub fn base(&self) -> usize {
+        self.base
+    }
+}
+
+
+pub struct PageManager<'a> {
+    metadata: &'a mut [PhysicalPage],
+    page_size: usize,
+}
+
+impl<'a> PageManager<'a> {
+
+    fn count_pages(regions: impl Iterator<Item = fdt::standard_nodes::MemoryRegion>, page_size: usize) -> usize {
+        regions
+            .map(|region| {
+                (region.starting_address, unsafe {
+                    region.starting_address.add(region.size.unwrap_or(0))
```

Comment:
Could we not unwrap only. If there is an error here, we should not continue any further

Reply:
It's not exactly an error.
Further regions could have a proper size.